File: csharp/account/domain/AccountService.cs
namespace ex.account;

using scyna;

public class AccountService
{
    public static IRepository loadRepository(Context context)
    {
        /*TODO inject loader function*/
        return Repository.load(context);
    }

    public static void AssureAccountNotExists(IRepository repository, EmailAddress email)
    {
        try
        {
            repository.GetAccountByEmail(email);
            throw Error.ACCOUNT_EXISTS;
        }
        catch (scyna.Error e)
        {
            if (e != Error.ACCOUNT_NOT_FOUND) throw;
        }
    }

}

File: go/account/domain/account_service.go
package domain

import (
	"ex/account/model"
	"ex/account/repository"

	scyna "github.com/scyna/core"
)

type accountService struct {
	Repository *repository.AccountRepository
	context    scyna.Context
}

func NewAccountService(context scyna.Context) *accountService {
	return &accountService{
		context:    context,
		Repository: &repository.AccountRepository{LOG: context},
	}
}

func (account *accountService) AssureAccountNotExists(email model.EmailAddress) scyna.Error {
	if _, err := account.Repository.GetAccountByEmail(email); err == nil {
		return model.ACCOUNT_EXISTED
	}
	return nil
}
